### PR TITLE
Get network stats by switching network namespace on newer Docker versions.

### LINKED ISSUE #822: Zero stats for Docker containers network/blkio
On Fedora 22 with docker-1.7.0-6.git74e7a7a.fc22.x86_64, all Docker containers report zero network traffic & blkio. Is there something that needs to be configured on the host to enable this reporting?

### PR DESCRIPTION
**Pull request by jimmidyson**
Fixes #822

### CODE REVIEW

**Comment by vishh** [comment]
ok to test

File: container/libcontainer/helpers.go [hunk 1]
```diff
@@ -93,14 +99,101 @@ func GetStats(cgroupManager cgroups.Manager, networkInterfaces []string) (*info.
...
+// Borrowed from libnetwork Stats - https://github.com/docker/libnetwork/blob/master/sandbox/interface_linux.go
+// In older kernels (like the one in Centos 6.6 distro) sysctl does not have netns support. Therefore
+// we cannot gather the statistics from /sys/class/net/<dev>/statistics/<counter> files. Per-netns stats
+// are naturally found in /proc/net/dev in kernels which support netns (ifconfig relyes on that).
+const (
+	netStatsFile = "/proc/net/dev"
+	base         = "[  ]*%s:([  ]+[0-9]+){16}"
+)
+
+func scanInterfaceStats(data, ifName string) (info.InterfaceStats, error) {
```
Comment (vishh) [inline container/libcontainer/helpers.go]: How about parsing the entire file once, and index it based on the Interface name?

**Comment by vishh** [comment]
Thanks for this PR @jimmidyson :100:

**Comment by vishh** [comment]
The e2e tests for network stats are failing.

**Comment by jimmidyson** [comment]
@vishh Looking back looks like the e2e tests have failed on most PRs recently... although I would have expected this PR to fix those failures :)

I have just tested it manually against coreos beta (the reported failure) & it works fine. I can't work out how to run the integration tests manually against a local VM - seems like the integration tests are coupled to GCE... not very friendly IMO. Any way to run these locally against a VM?

**Comment by vishh** [comment]
The integration tests are run against a few linux distros. Here are the
instruction for running it against your VM:
https://github.com/google/cadvisor/blob/master/docs/integration_testing.md#integration-testing-cadvisor

**Comment by jimmidyson** [comment]
Right:

"Today We only support remote execution in Google Compute Engine since that is where we run our continuous builds."

Working on making it a bit more flexible - would you be open to that? I don't have a GCE account to run integration tests myself.

**Comment by jimmidyson** [comment]
@vishh I've just run integration tests against coreos beta on GCE & it worked fine. Any chance you could retest?

File: container/libcontainer/helpers.go (hunk 1 above)
Comment (jimmidyson) [inline container/libcontainer/helpers.go]: Changed to that - good suggestion, thanks.

**Comment by vishh** [comment]
Works for me! Thanks @jimmidyson!! LGTM.

**Comment by jimmidyson** [comment]
Thanks @vishh.

Don't suppose there's any chance we could get a release cut with this fix to update deps in kubernetes & openshift?

**Comment by vishh** [comment]
We can update cadvisor deps right away in kubernetes. I am not sure when
the next Kubernetes release will happen though.

**Comment by jimmidyson** [comment]
Ah OK so no need to tag cadvisor first? Just use a non tagged commit in kubernetes godeps? As long as it's in next release whenever that is I'm happy.

**Comment by vishh** [comment]
I think tagging makes sense! Created a tag '0.16.0.1'.

**Comment by jimmidyson** [comment]
Great thanks!

**Comment by jimmidyson** [comment]
@vishh One thing I have noticed in further (post-merge ;)) testing is when running in a Docker container cadvisor needs to be in the hosts `pid` namespace (see https://docs.docker.com/reference/run/#pid-settings-pid) to be able to switch network namespaces based on pid of the docker containers. I don't see this as a problem but something that needs to be documented. Shall I do that?

**Comment by vishh** [comment]
Ah! Thats true. It cannot access the procfs of host. Do we have to
guarantee backwards compat with old kernels? Figuring out veth interfaces
and getting stats via sysfs seems to be way cheaper than switching
namespaces every second for every container.

**Comment by jimmidyson** [comment]
No we don't - this is how docker works, switching namespaces. Switching namespaces is cheap, no?

**Comment by jimmidyson** [comment]
This is the approach of libnetwork & how docker itself describes how to do high performance metrics collection - see https://docs.docker.com/articles/runmetrics/#tips-for-high-performance-metric-collection.

**Comment by vishh** [comment]
Hmm. If we were to be network setup agnostic, I agree that entering the
namespace might be the best strategy.

**Comment by jimmidyson** [comment]
I think we can use this technique for all versions of docker & get rid of all the veth handling code.

I'm not sure what you mean by network setup agnostic? Aren't we? Shouldn't we be?

**Comment by vishh** [comment]
Yes. I agree. What I meant by network agnostic was not having to know if
the underlying networking mechanism is veth or ipvlan, etc.

**Pull request merged**